documentation/projects/proposals/popularity_optimizations/20230420-implementation_plan_popularity_optimizations.md

```diff
@@ -0,0 +1,488 @@
+  -- paginate the results more efficiently than using OFFSET.
+  CREATE TEMP TABLE rows_to_update AS
+    SELECT ROW_NUMBER() over() row_id, identifier
+    FROM image
+    WHERE provider='flickr' AND updated_on < max_timestamp;
+  CREATE INDEX ON rows_to_update(row_id);
+
+  SELECT COUNT(*) INTO max_row FROM rows_to_update;
+
+  -- Loop over `rows_to_update` 10k rows at a time, updating scores.
+  -- Avoid deadlock using SKIP LOCKED
+  FOR i IN min_row..max_row BY rows_per_loop LOOP
+   UPDATE image
+   SET standardized_popularity = standardized_image_popularity(
+       image.provider,
+       image.meta_data
+   )
+   WHERE identifier in (
+       SELECT identifier FROM rows_to_update
+       WHERE row_id >= i AND row_id < i+rows_per_loop
+       FOR UPDATE SKIP LOCKED
+   );
+   -- COMMIT this batch
+   COMMIT;
+END LOOP;
+END;
+$$;
+```
+
+This is much less efficient than doing a single `UPDATE`, but has clear
+advantages:
+
+- It allows provider DAGs to continue running unimpeded.
+- The data refresh can run at any time, even while the refresh tasks are
+  running.
+- Because updates are committed incrementally, each batch of updated scores will
+  be 'live' on the image table as soon as they are committed. Rather than
+  waiting for all rows to be calculated before making the update, as much
+  up-to-date data as possible will be present each time the data refresh runs.
+
+### Remove popularity steps from the data_refresh DAGs
+
+We can simply remove the popularity tasks from the relevant data refresh DAG
+factories. See the `Parallelizable Streams` section for more information on the
+timing of this change.
+
+### Run the data refresh from the media tables
+
+Once the `image` and `audio` tables contain up-to-date standardized popularity
+data, we can update the data refresh to read from these tables rather than their
+respective materialized views. Because the schema of the media tables now
+matches what the ingestion server already expects, we can simply change the
+table name each place it is referenced:
+
+- [`_get_shared_cols`](https://github.com/WordPress/openverse/blob/003c0d3e5c2813b671120144756659775f3369f4/ingestion_server/ingestion_server/ingest.py#L75)
+- [`refresh_api_table`](https://github.com/WordPress/openverse/blob/003c0d3e5c2813b671120144756659775f3369f4/ingestion_server/ingestion_server/ingest.py#L310)
+- [`get_copy_data_query`](https://github.com/WordPress/openverse/blob/003c0d3e5c2813b671120144756659775f3369f4/ingestion_server/ingestion_server/queries.py#L206)
+
+## Dependencies
+
+### Tools & packages
+
+<!-- Describe any tools or packages which this work might be dependent on. If multiple options are available, try to list as many as are reasonable with your own recommendation. -->
+
+We should not need any additional dependencies.
+
+### Other projects or work
+
+<!-- Note any projects this plan is dependent on. -->
+
+This does not depend on any existing projects.
```

Comment:
Thinking about this more, it might be a good idea to link to the project for [Documenting Media Properties](https://github.com/WordPress/openverse/issues/412) since we'll want to make sure that we document changes (like the addition of the `standardized_popularity` column) in accordance with that work.

Reply:
I was thinking we could make the SQL file parser configurable to allow for parsing `view` if it's available, and only use `image` if not - of course if Documenting Media Properties is implemented before this project :)